lib/charms/mysql/v0/architecture.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Library to provide hardware architecture checks for VMs and K8s charms.
+
+The WrongArchitectureWarningCharm class is designed to be used alongside
+the is-wrong-architecture helper function, as follows:
+
+```python
+import sys
+
+from ops import main
+from charms.mysql.v0.architecture import WrongArchitectureWarningCharm, is_wrong_architecture
+
+if __name__ == "__main__":
+    if is_wrong_architecture():
+        main(WrongArchitectureWarningCharm)
+```
+"""
+
+import logging
+import os
+import sys
+
+from ops import BlockedStatus, CharmBase
+
+# The unique Charmhub library identifier, never change it
+LIBID = "827e04542dba4c2a93bdc70ae40afdb1"
+LIBAPI = 0
+LIBPATCH = 1
+
+PYDEPS = ["ops>=2.0.0"]
+
+
+logger = logging.getLogger(__name__)
+
+
+class WrongArchitectureWarningCharm(CharmBase):
+    """A fake charm class that only signals a wrong architecture deploy."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        hw_arch = os.uname().machine
+        self.unit.status = BlockedStatus(f"Error: Charm incompatible with {hw_arch} architecture")
+        sys.exit(0)
+
+
+def is_wrong_architecture() -> bool:
+    """Checks if charm was deployed on wrong architecture."""
+    manifest_path = f"{os.environ.get('CHARM_DIR')}/manifest.yaml"
+
+    if not os.path.exists(manifest_path):
+        logger.error("Cannot check architecture: manifest file not found in %s", manifest_path)
+        return False
+
+    with open(manifest_path, "r") as file:
+        manifest = file.read()
+
+    hw_arch = os.uname().machine
+    if ("amd64" in manifest and hw_arch == "x86_64") or (
+        "arm64" in manifest and hw_arch == "aarch64"
+    ):
```

Comment:
I SSH-ed into a mysql-operator running pod, and found the `manifest.yaml`. In principle, the change should be relatively easy to make, although it will make the new _architecture_ lib dependent on the `pyyaml` package.

Do we want that? or are we OK with a bit of un-safety? It seems PostgreSQL folks are fine with the implementation (see [code](https://github.com/canonical/postgresql-k8s-operator/blob/70a96ce33eb0de054b9297e3bd558f6ece54c0b0/src/arch_utils.py#L39-L41)).

Reply:
Done in commit https://github.com/canonical/mysql-operator/pull/560/commits/c2c5e99609e6dc01dad0b074a93c4f0819ee785a.